lib/providers/notification_service.dart

```diff
@@ -0,0 +1,126 @@
+import 'package:flutter_riverpod/flutter_riverpod.dart';
+
+import '../graphql/client.dart';
+import '../helpers/event_bus.dart';
+import '../models/events/app_ready.dart';
+import '../theme/bccm_typography.dart';
+
+final notificationServiceProvider = Provider<NotificationService>((ref) {
+  return NotificationService(ref: ref);
+});
+
+class NotificationService {
+  Ref ref;
+  String? fcmToken;
+  late StreamSubscription<AppReadyEvent> _appReadySubscription;
+  StreamSubscription<RemoteMessage>? _onMessageSubscription;
+  StreamSubscription<RemoteMessage>? _onMessageOpenedApp;
+  StreamSubscription<String>? _tokenSubscription;
+  ProviderSubscription<Settings>? _settingsSubscription;
+
+  NotificationService({required this.ref}) {
+    _appReadySubscription = globalEventBus.on<AppReadyEvent>().listen(_onAppReady);
+    ref.onDispose(dispose);
+  }
+
+  void dispose() {
+    _appReadySubscription.cancel();
+    _onMessageSubscription?.cancel();
+    _onMessageOpenedApp?.cancel();
+    _tokenSubscription?.cancel();
+    _settingsSubscription?.close();
+  }
+
+  /// Request permission and get token to start receiving push notifications
+  Future requestPermissionAndSetup() async {
+    var result = await FirebaseMessaging.instance.requestPermission();
+    print('NotificationStatus: ${result.authorizationStatus}');
+    FirebaseMessaging.instance.setForegroundNotificationPresentationOptions(alert: false, badge: false, sound: true);
+
+    final token = await FirebaseMessaging.instance.getToken();
+    if (token != null) {
+      await _setDeviceToken(token);
+    } else {
+      debugPrint('FirebaseMessaging.instance.getToken() returned null');
+    }
+    _setupTokenListeners();
+  }
+
+  void _setupTokenListeners() {
+    _tokenSubscription ??= FirebaseMessaging.instance.onTokenRefresh.listen(_setDeviceToken)
+      ..onError((err) {
+        print('error onTokenRefresh');
+      });
+    _settingsSubscription ??= ref.listen<Settings>(settingsProvider, (old, value) async {
+      if (old?.appLanguage != value.appLanguage) {
+        var token = await FirebaseMessaging.instance.getToken();
+        if (token == null) return;
+        _setDeviceToken(token);
+      }
+    });
+  }
+
+  void _onAppReady(AppReadyEvent _) async {
+    var initialMessage = await FirebaseMessaging.instance.getInitialMessage();
+    if (initialMessage != null) {
+      debugPrint('Handling initial notification message.');
+      _handleMessage(initialMessage, openedFromBackground: true);
+    }
+
+    _onMessageSubscription ??= FirebaseMessaging.onMessage.listen(_handleMessage);
+    _onMessageOpenedApp ??= FirebaseMessaging.onMessageOpenedApp.listen((message) {
+      debugPrint('_onMessageOpenedApp');
+      _handleMessage(message, openedFromBackground: true);
+    });
+    debugPrint('Notification listeners set up.');
+  }
+
+  Future _setDeviceToken(String token) async {
+    debugPrint('Sending FCM token. $token');
+    fcmToken = token;
```

Comment:
This variable is not used

Reply:
Not here, but it's a public field. It's used in `developer_options.dart`